src/mlpack/bindings/julia/julia_util.cpp

```diff
@@ -0,0 +1,497 @@
+{
+  data::DatasetInfo d(pointsAreRows ? cols : rows);
+  for (size_t i = 0; i < d.Dimensionality(); ++i)
+  {
+    d.Type(i) = (dimensions[i]) ? data::Datatype::categorical :
+        data::Datatype::numeric;
+  }
+
+  arma::mat m(memptr, rows, cols, false, true);
+  std::get<0>(CLI::GetParam<std::tuple<data::DatasetInfo, arma::mat>>(
+      paramName)) = std::move(d);
+  std::get<1>(CLI::GetParam<std::tuple<data::DatasetInfo, arma::mat>>(
+      paramName)) = pointsAreRows ? std::move(m.t()) : std::move(m);
+  CLI::SetPassed(paramName);
+}
+
+/**
+ * Call CLI::GetParam<int>().
+ */
+int CLI_GetParamInt(const char* paramName)
+{
+  return CLI::GetParam<int>(paramName);
+}
+
+/**
+ * Call CLI::GetParam<double>().
+ */
+double CLI_GetParamDouble(const char* paramName)
+{
+  return CLI::GetParam<double>(paramName);
+}
+
+/**
+ * Call CLI::GetParam<std::string>().
+ */
+const char* CLI_GetParamString(const char* paramName)
+{
+  return CLI::GetParam<std::string>(paramName).c_str();
+}
+
+/**
+ * Call CLI::GetParam<bool>().
+ */
+bool CLI_GetParamBool(const char* paramName)
+{
+  return CLI::GetParam<bool>(paramName);
+}
+
+/**
+ * Call CLI::GetParam<std::vector<std::string>>() and get the length of the
+ * vector.
+ */
+size_t CLI_GetParamVectorStrLen(const char* paramName)
+{
+  return CLI::GetParam<std::vector<std::string>>(paramName).size();
+}
+
+/**
+ * Call CLI::GetParam<std::vector<std::string>>() and get the i'th string.
+ */
+const char* CLI_GetParamVectorStrStr(const char* paramName, const int i)
+{
+  return CLI::GetParam<std::vector<std::string>>(paramName)[i].c_str();
+}
+
+/**
+ * Call CLI::GetParam<std::vector<int>>() and get the length of the vector.
+ */
+size_t CLI_GetParamVectorIntLen(const char* paramName)
+{
+  return CLI::GetParam<std::vector<int>>(paramName).size();
+}
+
+/**
+ * Call CLI::GetParam<std::vector<int>>() and return a pointer to the vector.
+ * The vector will be created in-place and it is expected that the calling
+ * function will take ownership.
+ */
+uint64_t* CLI_GetParamVectorIntPtr(const char* paramName)
+{
+  const size_t size = CLI::GetParam<std::vector<int>>(paramName).size();
+  uint64_t* ints = new uint64_t[size];
+
+  for (size_t i = 0; i < size; ++i)
+    ints[i] = CLI::GetParam<std::vector<int>>(paramName)[i];
+
+  return ints;
+}
+
+/**
+ * Get the number of rows in a matrix parameter.
+ */
+size_t CLI_GetParamMatRows(const char* paramName)
+{
+  return CLI::GetParam<arma::mat>(paramName).n_rows;
+}
+
+/**
+ * Get the number of columns in a matrix parameter.
+ */
+size_t CLI_GetParamMatCols(const char* paramName)
+{
+  return CLI::GetParam<arma::mat>(paramName).n_cols;
+}
+
+/**
+ * Get the memory pointer for a matrix parameter.
+ * Note that this will assume that whatever is calling will take ownership of
+ * the memory!
+ */
+double* CLI_GetParamMat(const char* paramName)
+{
+  arma::access::rw(CLI::GetParam<arma::mat>(paramName).mem_state) = 1;
```

Comment:
I am curious whether we should take into account preallocated storage in `arma::mat` like we do in `CLI_GetParamUMat` .

Reply:
Nice catch, you're right!  Fixed in 004e104be. :)